compiler/rustc_lint/src/types.rs

```diff
@@ -486,6 +508,62 @@ fn lint_literal<'tcx>(
     }
 }
 
+fn lint_nan<'tcx>(
+    cx: &LateContext<'tcx>,
+    e: &'tcx hir::Expr<'tcx>,
+    binop: hir::BinOp,
+    l: &'tcx hir::Expr<'tcx>,
+    r: &'tcx hir::Expr<'tcx>,
+) {
+    fn is_nan(cx: &LateContext<'_>, expr: &hir::Expr<'_>) -> bool {
+        let expr = expr.peel_blocks().peel_borrows();
+        match expr.kind {
+            ExprKind::Path(qpath) => {
+                let Some(def_id) = cx.typeck_results().qpath_res(&qpath, expr.hir_id).opt_def_id() else { return false; };
+
+                matches!(cx.tcx.get_diagnostic_name(def_id), Some(sym::f32_nan | sym::f64_nan))
+            }
+            _ => false,
+        }
+    }
+
+    let is_l_nan = is_nan(cx, l);
+    let is_r_nan = is_l_nan || is_nan(cx, r);
+
+    if is_l_nan || is_r_nan {
+        let can_span_suggest =
+            l.span.can_be_used_for_suggestions() && r.span.can_be_used_for_suggestions();
+        cx.emit_spanned_lint(
+            INVALID_NAN_COMPARISONS,
+            e.span,
+            match binop.node {
+                hir::BinOpKind::Eq | hir::BinOpKind::Ne => InvalidNanComparisons::EqNe {
+                    suggestion: if can_span_suggest {
+                        if is_l_nan {
+                            InvalidNanComparisonsSuggestion::Spanful {
+                                nan_plus_binop: l.span.until(r.span),
+                                float: r.span.shrink_to_hi(),
+                                neg: (binop.node == hir::BinOpKind::Ne)
+                                    .then(|| r.span.shrink_to_lo()),
+                            }
+                        } else {
+                            InvalidNanComparisonsSuggestion::Spanful {
+                                nan_plus_binop: l.span.shrink_to_hi().to(r.span),
+                                float: l.span.shrink_to_hi(),
+                                neg: (binop.node == hir::BinOpKind::Ne)
+                                    .then(|| l.span.shrink_to_lo()),
+                            }
+                        }
+                    } else {
+                        InvalidNanComparisonsSuggestion::Spanless
+                    },
+                },
+                _ => InvalidNanComparisons::LtLeGtGe,
+            },
```

Comment:
Could you make this a local?

Reply:
What about:
```
let lint = match binop.node {
    Eq | Ne if is_l_nan => { ... }
    Eq | Ne if is_r_nan => { ... }
    Lt | Le | Gt | Ge if is_l_nan || is_r_nan => { .. }
    _ => return,
};
cx.emit_spanned_lint(..., lint)
```

Reply:
That works.

*The code is unfortunately still highly dense, maybe de-duplicating even more could help.*